nshlib/nsh_netcmds.c

```diff
@@ -1092,95 +1132,98 @@ int cmd_arp(FAR struct nsh_vtbl_s *vtbl, int argc, FAR char **argv)
        * xx.xx.xx.xx xx:xx:xx:xx:xx:xx xxxxxxxx[xxxxxxxx]
        */
 
-      nsh_output(vtbl, "%-12s %-17s Last Access Time\n",
-                 "IP Address", "Ethernet Address");
+      nsh_output(vtbl, "%-12s %-17s %s\n",
+                 "IP Address", "Ethernet Address", "Interface");
 
       for (i = 0; i < nentries; i++)
         {
+          FAR struct sockaddr_in *addr;
           FAR uint8_t *ptr;
 
+          if (ifname != NULL &&
+              strcmp(ifname, (FAR char *)arptab[i].arp_dev) != 0)
+            {
+              continue;
+            }
+
           /* Convert the IPv4 address to a string */
 
-          ptr = (FAR uint8_t *)&arptab[i].at_ipaddr;
+          addr = (FAR struct sockaddr_in *)&arptab[i].arp_pa;
+          ptr = (FAR uint8_t *)&addr->sin_addr.s_addr;
           snprintf(ipaddr, 16, "%u.%u.%u.%u",
                    ptr[0], ptr[1], ptr[2], ptr[3]);
 
           /* Convert the MAC address string to a binary */
 
           snprintf(ethaddr, 24, "%02x:%02x:%02x:%02x:%02x:%02x",
-                   arptab[i].at_ethaddr.ether_addr_octet[0],
-                   arptab[i].at_ethaddr.ether_addr_octet[1],
-                   arptab[i].at_ethaddr.ether_addr_octet[2],
-                   arptab[i].at_ethaddr.ether_addr_octet[3],
-                   arptab[i].at_ethaddr.ether_addr_octet[4],
-                   arptab[i].at_ethaddr.ether_addr_octet[5]);
-
-#ifdef CONFIG_SYSTEM_TIME64
-          nsh_output(vtbl, "%-12s %-17s 0x%" PRIx64 "\n",
-                     ipaddr, ethaddr, (uint64_t)arptab[i].at_time);
-#else
-          nsh_output(vtbl, "%-12s %-17s 0x%" PRIx32 "\n",
-                     ipaddr, ethaddr, (uint32_t)arptab[i].at_time);
-#endif
+                   (uint8_t)arptab[i].arp_ha.sa_data[0],
+                   (uint8_t)arptab[i].arp_ha.sa_data[1],
+                   (uint8_t)arptab[i].arp_ha.sa_data[2],
+                   (uint8_t)arptab[i].arp_ha.sa_data[3],
+                   (uint8_t)arptab[i].arp_ha.sa_data[4],
+                   (uint8_t)arptab[i].arp_ha.sa_data[5]);
```

Comment:
I think cast is not needed for `%02x`?

Reply:
char may map to either signed char or unsigned char, so the uint8_t is required to avoid 0x80 extend to 0xffffff80 in first case.